OpenRA.Mods.Common/Traits/World/Pathfinder/Graph.cs

```diff
@@ -0,0 +1,197 @@
+		public int LaneBias { get; set; }
+		public bool InReverse { get; set; }
+		public IActor IgnoredActor { get; set; }
+
+		readonly CellConditions checkConditions;
+		readonly IMobileInfo mobileInfo;
+		CellLayer<CellInfo> cellInfo;
+
+		public const int InvalidNode = int.MaxValue;
+
+		public Graph(CellLayer<CellInfo> cellInfo, IMobileInfo mobileInfo, IActor actor, IWorld world, bool checkForBlocked)
+		{
+			this.cellInfo = cellInfo;
+			this.World = world;
+			this.mobileInfo = mobileInfo;
+			this.Actor = actor;
+			this.LaneBias = 1;
+			this.checkConditions = checkForBlocked ? CellConditions.TransientActors : CellConditions.None;
+		}
+
+		// Sets of neighbors for each incoming direction. These exclude the neighbors which are guaranteed
+		// to be reached more cheaply by a path through our parent cell which does not include the current cell.
+		// For horizontal/vertical directions, the set is the three cells 'ahead'. For diagonal directions, the set
+		// is the three cells ahead, plus the two cells to the side, which we cannot exclude without knowing if
+		// the cell directly between them and our parent is passable.
+		static readonly CVec[][] DirectedNeighbors = {
+			new[] { new CVec(-1, -1), new CVec(0, -1), new CVec(1, -1), new CVec(-1, 0), new CVec(-1, 1) },
+			new[] { new CVec(-1, -1), new CVec(0, -1), new CVec(1, -1) },
+			new[] { new CVec(-1, -1), new CVec(0, -1), new CVec(1, -1), new CVec(1, 0), new CVec(1, 1) },
+			new[] { new CVec(-1, -1), new CVec(-1, 0), new CVec(-1, 1) },
+			CVec.Directions,
+			new[] { new CVec(1, -1), new CVec(1, 0), new CVec(1, 1) },
+			new[] { new CVec(-1, -1), new CVec(-1, 0), new CVec(-1, 1), new CVec(0, 1), new CVec(1, 1) },
+			new[] { new CVec(-1, 1), new CVec(0, 1), new CVec(1, 1) },
+			new[] { new CVec(1, -1), new CVec(1, 0), new CVec(-1, 1), new CVec(0, 1), new CVec(1, 1) },
+		};
+
+		public ICollection<GraphConnection> GetConnections(CPos position)
+		{
+			var previousPos = this.cellInfo[position].PreviousPos;
+
+			var dx = position.X - previousPos.X;
+			var dy = position.Y - previousPos.Y;
+			var index = dy * 3 + dx + 4;
+
+			var validNeighbors = new LinkedList<GraphConnection>();
+			var directions = DirectedNeighbors[index];
+			for (int i = 0; i < directions.Length; i++)
+			{
+				var neighbor = position + directions[i];
+				var movementCost = GetCostToNode(neighbor, directions[i]);
+				if (movementCost != InvalidNode)
+					validNeighbors.AddLast(new GraphConnection(position, neighbor, movementCost));
+			}
+
+			return validNeighbors;
+		}
+
+		int GetCostToNode(CPos destNode, CVec direction)
+		{
+			int movementCost;
+			if (this.mobileInfo.CanEnterCell(
+				this.World as World,
+				this.Actor as Actor,
+				destNode,
+				out movementCost,
+				this.IgnoredActor as Actor,
+				checkConditions) && !(this.CustomBlock != null && this.CustomBlock(destNode)))
+			{
+				return CalculateCellCost(destNode, direction, movementCost);
+			}
```

Comment:
I wouldn't, because these casts are over an interface. I want to have the ability to provide alternative implementation of map such as mocks. Casting without the "as" will rise an exception. Otherwise, will return a null which, for unit testing, is completely valid.